app/src/backlight.c

```diff
@@ -0,0 +1,180 @@
+
+#include <zmk/activity.h>
+#include <zmk/backlight.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/activity_state_changed.h>
+#include <zmk/events/usb_conn_state_changed.h>
+#include <zmk/usb.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define BACKLIGHT_LEVELS CONFIG_ZMK_BACKLIGHT_LEVELS
+
+BUILD_ASSERT(DT_HAS_CHOSEN(zmk_backlight), "No backlight device found");
+
+static const struct device *const bl_leds = DEVICE_DT_GET(DT_CHOSEN(zmk_backlight));
+
+struct backlight_state {
+    uint8_t brightness;
+    bool on;
+};
+
+static struct backlight_state state;
+
+static int zmk_backlight_update() {
+    if (state.on) {
+        uint8_t percent = state.brightness * 100 / (BACKLIGHT_LEVELS - 1);
+        return led_set_brightness(bl_leds, 0, percent);
+    } else {
+        return led_off(bl_leds, 0);
+    }
+}
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+static int bl_settings_set(const char *name, size_t len, settings_read_cb read_cb, void *cb_arg) {
+    const char *next;
+
+    if (settings_name_steq(name, "state", &next) && !next) {
+        if (len != sizeof(state)) {
+            return -EINVAL;
+        }
+
+        int rc = read_cb(cb_arg, &state, sizeof(state));
+        if (rc < 0) {
+            return rc;
+        }
+
+        return zmk_backlight_update();
+    }
+
+    return -ENOENT;
+}
+
+static struct settings_handler bl_conf = {.name = "backlight", .h_set = bl_settings_set};
+
+static void zmk_backlight_save_state_work() {
+    settings_save_one("backlight/state", &state, sizeof(state));
+}
+
+static struct k_delayed_work backlight_save_work;
+#endif // IS_ENABLED(CONFIG_SETTINGS)
+
+static int zmk_backlight_save_state() {
+#if IS_ENABLED(CONFIG_SETTINGS)
+    k_delayed_work_cancel(&backlight_save_work);
+    return k_delayed_work_submit(&backlight_save_work, K_MSEC(CONFIG_ZMK_SETTINGS_SAVE_DEBOUNCE));
+#else
+    return 0;
+#endif
+}
+
+static int zmk_backlight_init(const struct device *_arg) {
+    if (!device_is_ready(bl_leds)) {
+        LOG_ERR("Backlight device \"%s\" is not ready", bl_leds->name);
+        return -ENODEV;
+    }
+
+    state = (struct backlight_state){.brightness = CONFIG_ZMK_BACKLIGHT_LEVEL_START - 1,
+                                     .on = IS_ENABLED(ZMK_BACKLIGHT_ON_START)};
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+    settings_subsys_init();
+
+    int err = settings_register(&bl_conf);
+    if (err) {
+        LOG_ERR("Failed to register the backlight settings handler (err %d)", err);
+        return err;
+    }
+
+    k_delayed_work_init(&backlight_save_work, zmk_backlight_save_state_work);
+
+    settings_load_subtree("backlight");
+#endif
+
+    return zmk_backlight_update();
+}
+
+static int zmk_backlight_set_on(bool on) {
+    state.on = on;
+
+    int rc = zmk_backlight_update();
+    if (rc != 0) {
+        return rc;
+    }
+
+    return zmk_backlight_save_state();
+}
+
+static int zmk_backlight_adjust_brt(int direction) {
+    int b = state.brightness + direction;
+    if (b < 1) {
```

Comment:
I think this logic might not do what you want for `|direction| > 1`

I suggest something more like
```c
const int new_brightness = state.brightness + direction
state.on = new_brightness > 0;
state.brightness = CLAMP(new_brightness, 1, BACKLIGHT_LEVELS);
```

Reply:
I think this works correctly. Can you give me an example?

Reply:
You would expect decrementing by one twice would give the same result as decrementing by two, but it might not.

Scenario 1:

1. Start with `on = true, brightness = 2`.
2. Call `zmk_backlight_adjust_brt(-1)`. Now `on = true, brightness = 1`.
3. Call `zmk_backlight_adjust_brt(-1)` again. Now `on = false, brightness = 1`.

Scenario 2:

1. Start with `on = true, brightness = 2`.
2. Call `zmk_backlight_adjust_brt(-2)`. Now `on = false, brightness = 2`.

Reply:
I don't think it's a problem. Also your suggestion doesn't work:

1. Start with `on = true, brightness = 1`
1. Call `zmk_backlight_dec()`: `on = false, brightness = 1`
1. Call `zmk_backlight_inc()`, expected: `on = true, brightness = 1`, actual: `on = true, brightness = 2`

Reply:
If we're going to provide an API that allows for increments larger than 1, I think it needs to correctly handle them.

Here are a couple ideas that would handle incrementing from off properly. I'm not sure what people would expect the behavior to be if brightness > 1, you turn the backlight off, and then you increment the brightness though.

It could revert to the previous brightness plus the increment:

```c
// Let brightness decrease to 0, but set it back to 1 if turning on and the brightness is 0.
int zmk_backlight_adjust_brt(int direction) {
  const int new_brightness = state.brightness + direction
  state.on = new_brightness > 0;
  state.brightness = CLAMP(new_brightness, 0, BACKLIGHT_LEVELS);
  ...
}

int zmk_backlight_set_on(bool on) {
  state.on = on;
  if (on && state.brightness == 0) {
    state.brightness = 1;
  }
}
```

It could increment from zero:

```c
// Keep brightness >= 1, but if backlight is off, treat brightness as if it were 0.
int zmk_backlight_adjust_brt(int direction) {
  const int current_brightness = state.on ? state.brightness : 0;
  const int new_brightness = current_brightness + direction
  state.on = new_brightness > 0;
  state.brightness = CLAMP(new_brightness, 1, BACKLIGHT_LEVELS);
  ...
}
```

It could also just revert to the previous brightness without incrementing, then increment on the following change. I think if this was the desired behavior, we would want to handle that at a higher level though, as baking that into the adjust API would be confusing. 

```c
case BL_INC:
  if (zmk_backlight_is_on()) {
    zmk_backlight_inc();
  } else {
    zmk_backlight_set_on(true);
  }
```

Reply:
Given the chosen meaning of BACKLIGHT_LEVELS to include zero, things should be clamped to BACKLIGHT_LEVELS - 1 in the above examples.

Or if we were to allow setting arbitrary brightness, those examples would look more like this:

```c
// Allow brightness = 0
int zmk_backlight_get_brt(void) { return state.brightness; }
int zmk_backlight_set_brt(int brightness) {
  state.on = brightness > 0;
  state.brightness = CLAMP(brightness, 0, BACKLIGHT_LEVELS - 1);

  return zmk_backlight_update();
}

int zmk_backlight_adjust_brt(int direction) {
  return zmk_backlight_set_brt(zmk_backlight_get_brt() + direction);
}
```

```c
// Keep brightness >= 1
int zmk_backlight_get_brt(void) {
  return state.on ? state.brightness : 0; 
}

int zmk_backlight_set_brt(int brightness) {
  state.on = brightness > 0;
  state.brightness = CLAMP(brightness, 1, BACKLIGHT_LEVELS - 1);

  return zmk_backlight_update();
}

int zmk_backlight_adjust_brt(int direction) {
  return zmk_backlight_set_brt(zmk_backlight_get_brt() + direction);
}
```